src/engines/caching.cc

```diff
@@ -0,0 +1,315 @@
+#include "vcmap.h"
+#include <lib_acl.hpp>
+#include <lib_acl.h>
+#include <stream/socket_stream.hpp>
+#include <redis/redis_stream.hpp>
+#include <redis/redis_client.hpp>
+#include <redis/redis_result.hpp>
+#include <list>
+#include <libmemcached/memcached.h>
+#include <string>
+
+#define DO_LOG 0 
+#define LOG(msg) if (DO_LOG) std::cout << "[caching] " << msg << "\n"
+#define ZERO 0
+
+namespace pmemkv {
+namespace caching {
+
+CachingEngine::CachingEngine(const string& config) {
+    if (!readConfig(config) || !(basePtr = KVEngine::Start(subEngine, subEngineConfig)))
+       throw "CachingEngine Exception"; 
+    LOG("CachingEngine Started....");
+}
+
+CachingEngine::~CachingEngine() {
+    LOG("CachingEngine Destructor");
+    if (basePtr) {
+       KVEngine::Stop(basePtr);
+       LOG("CachingEngine Stopped....");
+    }
+}
+
+bool CachingEngine::readConfig(const string& config) {
+    bool retVal = false;
+    rapidjson::Document d;
+    if (d.Parse(config.c_str()).HasParseError()) return retVal;
+    /* Checking the mandatory fields from the config */
+    if (!d.HasMember("subengine") || !d["subengine"].IsString()) return retVal;
+    if (!d.HasMember("subengine_config") || !d["subengine_config"].IsObject()) return retVal;
+    if (!d["subengine_config"].HasMember("path") || !d["subengine_config"]["path"].IsString()) return retVal;
+    if (!d.HasMember("host") || !d["host"].IsString()) return retVal;
+    if (!d.HasMember("port") || !d["port"].IsInt()) return retVal;
+    if (!d.HasMember("remote_type") || !d["remote_type"].IsString()) return retVal;
+    if (!d.HasMember("remote_user") || !d["remote_user"].IsString()) return retVal;
+    if (!d.HasMember("remote_pwd") || !d["remote_pwd"].IsString()) return retVal;
+    if (!d.HasMember("remote_url") || !d["remote_url"].IsString()) return retVal;
+    if (!d.HasMember("attempts") || !d["attempts"].IsInt()) return retVal;
+    if (d.HasMember("ttl") && !d["ttl"].IsUint()) return retVal;
+
+    retVal = true;
+    rapidjson::StringBuffer sb;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+    d["subengine_config"].Accept(writer);
+    subEngineConfig = sb.GetString();
+    subEngine     = d["subengine"].GetString();
+    remoteType    = d["remote_type"].GetString();
+    remoteUser    = d["remote_user"].GetString();
+    remotePasswd  = d["remote_pwd"].GetString();
+    remoteUrl     = d["remote_url"].GetString();
+    ttl           = d.HasMember("ttl") ? d["ttl"].GetInt() : 0;
+    host          = d["host"].GetString();
+    port          = d["port"].GetInt();
+    attempts      = d["attempts"].GetInt();
+    retVal        = true;
+    return retVal;
+}
+
+void CachingEngine::All(void* context, KVAllCallback* callback) {
+    LOG("All");
+    string result;
+    Each(&result, [](void* context, int kb, const char* k, int vb, const char* v) {
+        const auto c = ((string*) context);
+        c->append(string(k, kb));
+        c->append(string(v, vb));
+      });
```

Comment:
Variable `result` should be a counter here, not a string.  This code will attempt to build a string in DRAM using the entire contents of what's currently in the cache, which can be very large.  This code will result in out-of-memory conditions, without any benefit since the contents of the result string are never used.  Much better for result to be a counter that is incremented by the callback, and then the subsequent `result.empty()` check turns into `result > 0`.